### PR TITLE
interpolator support in Cube

### PR DESCRIPTION
**Pull request by gsps1**
Follow up on #1251, adresses:

interpolators

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesInterpolator.java [hunk 1]
```diff
@@ -0,0 +1,139 @@
+/*
+ * Copyright © 2014 Cask Data, Inc.
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesInterpolator.java]: 2015

Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesInterpolator.java]: update your IDEA settings ;)

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesInterpolator.java [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+import com.google.common.collect.AbstractIterator;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterators;
+import com.google.common.collect.PeekingIterator;
+
+import java.util.Collection;
+import java.util.Iterator;
+
+/**
+ * Given a timeseries, interpolates the values at each timestamp between the earliest and
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesInterpolator.java]: Applies given interpolator to a given time series. Your text goes after that

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/CubeQuery.java [hunk 3]
```diff
@@ -50,17 +51,19 @@
...
     this(startTs, endTs, resolution, -1,
          measureName, measureType,
-         sliceByTagValues, groupByTags);
+         sliceByTagValues, groupByTags, null);
   }
 
   public CubeQuery(long startTs, long endTs, int resolution, int limit,
                    String measureName, MeasureType measureType,
-                   Map<String, String> sliceByTagValues, List<String> groupByTags) {
+                   Map<String, String> sliceByTagValues, List<String> groupByTags, Interpolator interpolator) {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/CubeQuery.java]: @ Nullable

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java [hunk 4]
```diff
@@ -16,24 +16,27 @@
...
-import com.google.common.collect.Lists;
-import com.google.common.collect.Maps;
-import com.google.common.collect.Table;
-
-import java.util.Collection;
-import java.util.Collections;
-import java.util.List;
-import java.util.Map;
-import javax.annotation.Nullable;
+  import co.cask.cdap.metrics.data.TimeSeriesInterpolator;
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: why spaces before import?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java [hunk 5]
```diff
@@ -231,15 +234,31 @@ private Aggregation findAggregation(CubeQuery query) {
...
+    }
     for (Map.Entry<Map<String, String>, Map<Long, Long>> row : aggValuesToTimeValues.rowMap().entrySet()) {
       List<TimeValue> timeValues = Lists.newArrayList();
       for (Map.Entry<Long, Long> timeValue : row.getValue().entrySet()) {
         timeValues.add(new TimeValue(timeValue.getKey(), timeValue.getValue()));
       }
       Collections.sort(timeValues);
-      result.add(new TimeSeries(query.getMeasureName(), row.getKey(), timeValues));
+      PeekingIterator<TimeValue> timeValueItor = Iterators.peekingIterator(
+        new TimeSeriesInterpolator(timeValues, query.getInterpolator(), query.getResolution()).iterator());
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: can query.getInterpolator() be null

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java [hunk 6]
```diff
@@ -231,15 +234,31 @@ private Aggregation findAggregation(CubeQuery query) {
...
+      long resultTimeStamp = query.getStartTs();
+      List<TimeValue> resultTimeValues = Lists.newArrayList();
+      for (int i = 0; i < count; i++) {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: don't fully get the logic. I though all this done in TimeSeriesInterpolator? looping and filling zeroes or interpolated values

**Comment by abaranau** [comment]
couple comments - biggest one is about same logic in DefaultCube and TimeSeriesInterpolator

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java (hunk 5 above)
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: yes

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java (hunk 6 above)
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/store/cube/DefaultCube.java]: when the interpolator is null, we might not have values for all the data points and this adds them. as we discussed yesterday, we can have a default interpolator that fills in 0's and in that case we would loop but we the zero's will be supplied by the interpolator.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesInterpolator.java [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+ *   5   -   -   -   3
+ *   5   5   5   5   3
+ */
+public class TimeSeriesInterpolator implements Iterable<TimeValue> {
+
+  private final Collection<TimeValue> timeSeries;
+  private final Interpolator interpolator;
+  private final int resolution;
+
+  public TimeSeriesInterpolator(Collection<TimeValue> timeValues, Interpolator interpolator, int resolution) {
```
Comment (abaranau) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/data/TimeSeriesInterpolator.java]: @ Nullable interpolator? same at field level.

**Comment by abaranau** [comment]
LGTM

**Comment by abaranau** [comment]
that small comment only

**Pull request merged**